external/gtest/CMakeLists.txt

```diff
@@ -1,43 +1,41 @@
 cmake_minimum_required(VERSION 3.10)
 project(GoogleTest CXX)
 
+# Warning fixes - borrowed from GTest build scripts
 if(WIN32)
-    set(gtest_cxx_flags -D_UNICODE -DUNICODE -DSTRICT -DWIN32_LEAN_AND_MEAN)
+  set(gtest_cxx_flags -D_UNICODE -DUNICODE -DSTRICT -DWIN32_LEAN_AND_MEAN)
 else()
-    set(gtest_cxx_flags -Wno-unused-parameter -Wno-missing-field-initializers)
+  set(gtest_cxx_flags -Wno-unused-parameter -Wno-missing-field-initializers)
 endif()
 
 # gtest
 
-add_library(gtest STATIC EXCLUDE_FROM_ALL
-    "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/src/gtest-all.cc"
+add_library(gtest STATIC
+  "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/src/gtest-all.cc"
 )
 
 target_include_directories(gtest
-    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest"
-    PUBLIC "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/include"
+  PRIVATE "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest"
+  PUBLIC "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/include"
 )
 
-target_compile_options(gtest
-    PRIVATE ${gtest_cxx_flags}
+target_compile_options(gtest PRIVATE ${gtest_cxx_flags}
 )
 
 if (UNIX)
-    target_link_libraries(gtest PRIVATE pthread)
+  target_link_libraries(gtest PRIVATE pthread)
 endif(UNIX)
 
 # gtest_main
 
-add_library(gtest_main STATIC EXCLUDE_FROM_ALL
-    "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/src/gtest_main.cc"
+add_library(gtest_main STATIC
+  "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/src/gtest_main.cc"
 )
 
 target_include_directories(gtest_main
-    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/include"
+  PRIVATE "${CMAKE_CURRENT_LIST_DIR}/googletest/googletest/include"
```

Comment:
why `CMAKE_CURRENT_LIST_DIR` instead of less tricky and common known `CMAKE_CURRENT_SOURCE_DIR`?

Reply:
Because they are way different.

If you'd use `include(gtest/CMakeLists.txt)` (just because "why not?") instead of `add_subdirectory(gtest)`, the build will fail.

How is `CMAKE_CURRENT_LIST_DIR` tricky?

Reply:
> If you'd use include(gtest/CMakeLists.txt) (just because "why not?") instead of add_subdirectory(gtest), the build will fail.

When it is the case, we immediately will use `CMAKE_CURRENT_LIST_DIR`. But for now `CMAKE_CURRENT_SOURCE_DIR` is enough. I'd like changes that by itself mean what is enough, especially here, in the project of _educational purpose_ .

Reply:
I have to disagree.

Using `CMAKE_CURRENT_SOURCE_DIR ` to specify a relative to current _CMakeLists.txt_ file location - is a **bug**. You are relying on the behavior that happens to work fine in current particular case, but the documentation is clear that it is not always will be.

That's exactly the purpose of `CMAKE_CURRENT_LIST_DIR` variable, to use in cases like this.

The code is perfectly clear and exact.

Reply:
> less tricky and common known

Not sure why do you think so, but it feels like that knowledge come not from a pleasant experience.

Check out [this article](https://cgold.readthedocs.io/en/latest/tutorials/cmake-sources/common.html). 
If you want to skip heavy reading - there is short [Recommendation](https://cgold.readthedocs.io/en/latest/tutorials/cmake-sources/common.html#recommendation) section.

Reply:
Found another [example](https://crascit.com/2016/01/31/enhanced-source-file-handling-with-target_sources/) from one of CMake's maintainers.

Reply:
OK, I find separate PR more than acceptable.